libs/utils/WordPressUtils/src/main/java/org/wordpress/android/util/MediaUtils.java

```diff
@@ -386,20 +387,145 @@ public static String getExtensionForMimeType(String mimeType) {
         return fileExtensionFromMimeType.toLowerCase();
     }
 
-    public static String getPathFromContentUri(Context context, Uri imageUri) {
-        if(context == null || imageUri == null) {
-            return null;
+    /**
+     * Get a file path from a Uri. This will get the the path for Storage Access
+     * Framework Documents, as well as the _data field for the MediaStore and
+     * other file-based ContentProviders.
+     *
+     * Based on paulburke's solution for aFileChooser - https://github.com/iPaulPro/aFileChooser
+     *
+     * @param context The context.
+     * @param uri The Uri to query.
+     */
+    public static String getPath(final Context context, final Uri uri) {
+        String path = getDocumentProviderPathKitkatOrHigher(context, uri);
+
+        if (path!= null) {
+            return path;
         }
 
-        String path = null;
-        String[] projection = new String[]{MediaStore.Images.Media.DATA};
-        Cursor cur = context.getContentResolver().query(imageUri, projection, null, null, null);
-        if (cur != null && cur.moveToFirst()) {
-            int dataColumn = cur.getColumnIndex(MediaStore.Images.Media.DATA);
-            path = cur.getString(dataColumn);
+        // MediaStore (and general)
+        if ("content".equalsIgnoreCase(uri.getScheme())) {
+            return getDataColumn(context, uri, null, null);
+        }
+        // File
+        else if ("file".equalsIgnoreCase(uri.getScheme())) {
+            return uri.getPath();
         }
-        SqlUtils.closeCursor(cur);
-        return path;
+
+        return null;
+    }
+
+    @TargetApi(Build.VERSION_CODES.KITKAT)
+    private static String getDocumentProviderPathKitkatOrHigher(final Context context, final Uri uri) {
+        final boolean isKitKat = Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT;
+
+        // DocumentProvider
+        if (isKitKat && DocumentsContract.isDocumentUri(context, uri)) {
+            // ExternalStorageProvider
+            if (isExternalStorageDocument(uri)) {
+                final String docId = DocumentsContract.getDocumentId(uri);
+                final String[] split = docId.split(":");
+                final String type = split[0];
+
+                if ("primary".equalsIgnoreCase(type)) {
+                    return Environment.getExternalStorageDirectory() + "/" + split[1];
+                }
+
+                // TODO handle non-primary volumes
+            }
+            // DownloadsProvider
+            else if (isDownloadsDocument(uri)) {
+
+                final String id = DocumentsContract.getDocumentId(uri);
+                final Uri contentUri = ContentUris.withAppendedId(
```

Comment:
Is there some way to avoid hard-coding the downloads path here? This seems very fragile.